utils/pattern/src/parser/mod.rs

```diff
@@ -0,0 +1,315 @@
+/// than implementing a reversed infallible iterator (where the [`Item`] would be
+/// `Option<Result<Item>>`).
+///
+/// That decision impacts the ergonomics of operating on the parser, on one hand making
+/// the fallible iteration more ergonomic, at a trade-off of losing access to the wide
+/// range of Rust iterator traits.
+///
+/// ## Generic Placeholder
+///
+/// To handle generic placeholder design, the only constrain necessary in the parser
+/// is that a placeholder must be parsed from a string slice.
+/// At the moment of writing, Rust is [preparing to deprecate][`RFC 2924`] [`FromStr`] in favor of
+/// [`TryFrom<&str>`][`TryFrom`].
+/// Among many banfits of such transition would be the auto-trait behavior of [`From`] and
+/// a [`TryFrom<&str>`][`TryFrom`] for [`&str`] allowing for placeholders to be [`&str`] themselves.
+///
+/// Unfortunately, at the moment [`TryFrom<&str>`][`TryFrom`] for [`usize`] is not implemented, which would
+/// impact the core use case of placeholder patterns.
+///
+/// In result, the decision has been made to use [`FromStr`] for the time being, until
+/// [`TryFrom<&str>`][`TryFrom`] gets implemented on all types that support [`FromStr`].
+///
+/// [`TR35 2.6.1]: https://unicode.org/reports/tr35/tr35-dates.html#dateTimeFormat
+/// [`RFC 2924`]: https://github.com/rust-lang/rfcs/pull/2924
+/// [`Item`]: std::iter::Iterator::Item
+/// [`TryFrom`]: std::convert::TryFrom
+pub struct Parser<'p> {
+    input: &'p str,
+    len: usize,
+
+    start_idx: usize,
+    idx: usize,
+
+    state: ParserState,
+}
+
+impl<'p> Parser<'p> {
+    /// Creates a new `Parser`.
+    ///
+    /// # Examples
+    /// ```
+    /// use icu_pattern::Parser;
+    /// let mut parser = Parser::new("{0}, {1}");
+    /// ```
+    pub fn new(input: &'p str) -> Self {
+        Self {
+            input,
+            len: input.len(),
+
+            start_idx: 0,
+            idx: 0,
+
+            state: ParserState::default(),
+        }
+    }
+
+    /// An iterator method that advances the iterator and returns the result of an attempt to parse
+    /// the next token.
+    ///
+    /// # Examples
+    /// ```
+    /// use icu_pattern::{Parser, PatternToken};
+    ///
+    /// let mut parser = Parser::new("{0}, {1}");
+    ///
+    /// // A call to try_next() returns the next value…
+    /// assert_eq!(Ok(Some(PatternToken::Placeholder(0))), parser.try_next());
+    /// assert_eq!(Ok(Some(PatternToken::Literal { content: ", ", quoted: false})), parser.try_next::<usize>());
+    /// assert_eq!(Ok(Some(PatternToken::Placeholder(1))), parser.try_next());
+    ///
+    /// // … and then None once it's over.
+    /// assert_eq!(Ok(None), parser.try_next::<usize>());
+    /// ```
+    pub fn try_next<P>(
+        &mut self,
+    ) -> Result<Option<PatternToken<'p, P>>, ParserError<<P as FromStr>::Err>>
+    where
+        P: FromStr,
+        P::Err: Debug,
+    {
+        while let Some(b) = self.input.as_bytes().get(self.idx) {
```

Comment:
Issue: We've had pretty embarrassing problems (#503) by using bytes instead of chars to loop over a string.  Please use characters instead of bytes unless you're 100% sure that the input is ASCII-only (such as a TinyStr).

Related: Add tests for patterns that include non-ASCII characters.

Reply:
That's a bit of leftover from using `Slice` but I'm also generally against that. I disagree that we had "pretty embarrassing problems" - we had one, and I don't consider it embarrassing.

While I understand the incentive to use chars, the value of using bytes is that we're getting zero-overhead peekable iteration by being able to treat string slice as an array with an index and moving the index back, forth and skipping as we like at O(1). Operating on `chars` is more costly and limiting, and I treat it as an optional optimization once the algorithms are complete.

I'm happy to debate it further but I'd prefer this to be the last point of contention in this review and fix everything else first.

> Related: Add tests for patterns that include non-ASCII characters.

Yup, will port all the tests from DTF that include non-ASCII characters.

Reply:
> I treat [operating on chars] as an optional optimization once the algorithms are complete.

I disagree at a fundamental level with that point of view.  Step 1 should always be to write *correct* code; step 2 can be to optimize that code if we can prove certain invariants to be true.  Using chars is correct; using bytes is only correct under certain conditions, such as when the string is ASCII-only.  So we should always start with chars, never bytes.

> While I understand the incentive to use chars, the value of using bytes is that we're getting zero-overhead peekable iteration by being able to treat string slice as an array with an index and moving the index back, forth and skipping as we like at O(1). Operating on chars is more costly and limiting

I acknowledge that bytes are more convenient, but that's not an excuse to write incorrect code.

Reply:
> I disagree at a fundamental level with that point of view.

Well, we have a disagreement on a fundamental level then :)

> I acknowledge that bytes are more convenient, but that's not an excuse to write incorrect code.

What you and the other person define as a mean to achieve correct code may differ.
I don't think it's an acceptable way to communicate to imply that your interlocutor is looking for an excuse to write incorrect code.

Reply:
If you consider using bytes only as a means to an end, and you plan to fix it before checking in this PR, that's fine by my standards; just leave a TODO in the code acknowledging that.

As a code reviewer, it is my duty to flag areas where code is incorrect.  *If we now agree that the code you wrote is incorrect*, then the response I expect from the author of that code is to acknowledge that it is incorrect and commit to fixing it before merging the PR, not to dismiss the problem I found as an "optional optimization" and a "debate" to be had later.

Reply:
I guess the problem is that I disagree with your assessment. I think it's something we should discuss :)

Reply:
Also, you implied that i dismissed a problem you found. I disagree that I did. I am not dismissing your position just by disagree with with it.

Reply:
Will discuss further offline. Sorry for the use of strong language earlier.